### PR TITLE
Crash reporting

### PR DESCRIPTION
**Pull request by s9ferech**
### Pull Request Description

This is meant to solve issue #808. When a Rust panic occurs on the front-end then we reload the browser window and offer the user to report it to a server.

Crash-Reporting

Explanation of the GIF: The logging server is running in the background. The IDE is started with a debug scene that immediately causes a panic. We detect the panic and reload the whole GUI (which isn't really visible since there was is nothing in the window). On restart, the IDE shows a crash notification and allows us to report the stack trace to the server which writes it to a file.

### Important Notes

 We use a custom panic hook to throw JavaScript exceptions on panics. Those exceptions are caught through a global error handler. In addition to Rust panics, we also handle uncaught JavaScript errors and rejected promises. The latter might be too restrictive.

The reported stack trace contains multiple frames that are not relevant, but there does not seem to be a clean way to filter them out.

The implementation of the server can be found in the "src/js/logging_service" directory and can be run through `npm run start`. It is not included in the build system yet. The server takes http POST requests and writes their body to a new file in the "log" subdirectory.

The content security policy of the front-end needed to be relaxed to include `connect-src ... http://localhost:* ...;`.
To avoid security risks, the server can currently only take requests from localhost and only if the IDE is also running through localhost.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md) style guide.
- [ ] All code has automatic tests where possible.
- [ ] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [ ] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

**Review by wdanilo — CHANGES_REQUESTED** [review]
Hi @s9ferech, good to see the PR ❤️ 

I'd ask you for a few improvemments before getting into full review process here:
1. Instead of creating `src/js/logging_service`, create `src/js/lib/logging` and use `lerna` to manage it, as other things in `src/js`. This way it will be more uniform and you'd not need to commit `+1251` lines of `package-lock.json` which is already committed in parent dir.
2. Please be sure you have tested the required things and mark it in the PR description. This will save us a ton of time:
    image
3. Could you add some automatic tests to check whether this functionality works correctly? As this is very rarely seen in GUI, testing it properly would be important. I'm not yet sure how to design these tests here, so this is an open question for now - before you implement tests, please write your ideas how they could be implemented and let's discuss it.
4. Could you create a small giff showing this system in action and could you paste it in the PR? We always do it as it shows everyone what the PR is about :)

**Comment by s9ferech** [comment]
I think for testing, we could use a debug scene that immediately makes the IDE crash. Then we could launch it Spectron and inspect it from the outside and simulate the reporting.

**Comment by s9ferech** [comment]
I did try the test scenarios, but on my machine, the IDE always behaved quite differently to the description in the scenarios. For example, I could never see any visualisations. I am pretty sure that my changes couldn't break anything on that level though

**Comment by wdanilo** [comment]
@s9ferech I think that using spectron now would be overkill, but definitely, a debug scene that shows it working will be cool and we can stay with it for now!

Regarding no visualizations - are you running enso engine service next to GUI? Without it, you cannot get visualizations (in our dev guide there are sentences about it)

**Comment by s9ferech** [comment]
The debug scene is implemented already. The name is just "panic".

I am running the project manager. Otherwise, the IDE wouldn't even start in standard mode, I think.

File: src/js/lib/logging/server.js
Comment (wdanilo) [inline src/js/lib/logging/server.js:25]: This literal is used in 4 different places across different files. Please use literals only once, always.

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [inline src/js/lib/logging/test/test.js:35]: The name of this function does not really tell what does it do. The names should be more semantic.

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [inline src/js/lib/logging/test/test.js:37]: I dont understand this comment tbh. What does it mean that "for some reason we load a file"?

File: src/rust/lib/system/web/src/lib.rs
Comment (wdanilo) [inline src/rust/lib/system/web/src/lib.rs:586]: I'd use here inline JS instead. This would be easier for the reader to track whats going on, especially as the JS file is so small. But Im also completel ok with how it looks now, so its up to your preference :)

**Comment by wdanilo** [comment]
@s9ferech I have to admit I'm very nicely surprised of the overall code shape here. I'm super happy seeing it in such a good shape, really amazing work ❤️

File: src/rust/lib/system/web/src/lib.rs
Comment (s9ferech) [reply src/rust/lib/system/web/src/lib.rs:586]: I do prefer it in a separate file because it gives you IDE support. Inline, I don't even get syntax highlighting.

File: src/js/lib/content/src/index.js
Comment (wdanilo) [reply src/js/lib/content/src/index.js:151]: @s9ferech Could you elaborate more? First of all, I did not know that there is runtime difference between these two. Secondly, if this difference exists there and we have "slippery" situation of handlers that may not be run because of window refresh, this should be relaly well documented in the code, so no one will refactor it in the future in wrong way

File: src/js/lib/content/src/index.js
Comment (wdanilo) [reply src/js/lib/content/src/index.js:168]: We want the GUI to feel bulletproof. Basically if something crashes and we are able to automatically rescue, its always better than telling user "press here to try to refresh the gui". Regarding crash on closing the gui – we should not care , as closing the gui happens when closing the tab, so even if crash occurs, the tab will be just closed, right?

File: src/js/lib/logging/README.md
Comment (wdanilo) [reply src/js/lib/logging/README.md:3]: Why do you need this port number in `content.js`? Help info is provided by `client.js` and also content server is run from there? Regarding the log server - that's a very valid point. In such a case I would store it in some config, like `src/js/config.js` which will be imported by every package which needs it. (of course, add good docs to config.js and the variables stored there)

File: src/js/lib/logging/README.md
Comment (wdanilo) [reply src/js/lib/logging/README.md:1]: @s9ferech Why not? I mean, I dont see any downsides of keeping it there, but maybe I missing something - if so, I'd love to hear more abotu why you feel its not the proper place for it! :)

File: src/js/lib/logging/server.js
Comment (wdanilo) [reply src/js/lib/logging/server.js:10]: In the case of this file, I would create following sections: `Cmd Args`, `Server`, and `File Utils`. Basically, sections correspond to a logical division of the file which may actually become "split points" in the future. If you'd like to split this file into subsequent modules, I believe that server implementation could go to one file, arg parsing to another, and all utils for storing fiels and date mangling to yet another. But as they are so short right now, lets just keep them in a single file, but jsut add sections. Does it make sense to you?

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [reply src/js/lib/logging/test/test.js:1]: Our CI config is at `.github/workflows/check.yml`. I was never creating workflows for JS tests, so I cant help more with it now. In case you'd not be able to do it fast, lets leave it without this connection and lets do it as a separate pr some day in the future.

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [reply src/js/lib/logging/test/test.js:35]: Ah, this is inside of tests, ignore me here, sorry! It's ok!

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [reply src/js/lib/logging/test/test.js:37]: `some module that the server uses loads such a file at runtime` - this is like internally requiredd by express or is this a file that we require? It seems very strange to e that server tries to load a file during runtime implicitly.

File: src/rust/lib/system/web/src/lib.rs
Comment (wdanilo) [reply src/rust/lib/system/web/src/lib.rs:585]: We are putting `#[allow(unsafe_code)]` above extern "C" functions in such scenarions. Thsi praga tells just that you as the author, fully approves the usage of unsafe code in this place.

**Comment by wdanilo** [comment]
@s9ferech I answered your questions. Ping me whenever you'll think its ready for re-review :)

File: src/js/lib/logging/test/test.js
Comment (s9ferech) [reply src/js/lib/logging/test/test.js:37]: I don't understand exactly why this happens. As far as I can tell, some code that is called by express dynamically loads a module at runtime. Our code is not responsible for this.

File: src/js/lib/content/src/index.js
Comment (s9ferech) [reply src/js/lib/content/src/index.js:151]: I just tested this and it turns out that my assumptions were wrong. This assignment seems to be equivalent to `addEventListener` indeed, except that it can be overwritten.

As far as I know, we only use one single listener on this event at the moment, and it seems very unlikely that we would add another one. For that reason, I think that it is no big problem anyway.

File: src/js/lib/content/src/index.js
Comment (s9ferech) [reply src/js/lib/content/src/index.js:168]: We don't prevent the user from closing the tab. That is right.

File: src/js/lib/logging/README.md
Comment (s9ferech) [reply src/js/lib/logging/README.md:3]: The `client.js` currently passes the logging-server address to the window as a URL parameter. If you open the GUI by hand in an external web browser and don't submit this parameter then we need to get the address from somewhere else. This is why `content.js` also needs access to the default.

Comment (s9ferech) [reply src/js/lib/logging/README.md:3]: The term config (or rather `cfg`) is already used in different places for different things, like `urlCfg`, `serverCfg`, `windowCfg` or just `cfg`. If it seems likely that we will add more to this one then i would refer to it as just `cfg` and rename the existing `cfg` in `content/index.js` to `urlCfg` which would be consistent with `client/index.js`. Is that fine or do you prefer another solution?

File: src/js/lib/logging/README.md
Comment (s9ferech) [reply src/js/lib/logging/README.md:1]: Because it is currently not intended as a library but rather as a stand-alone executable.

File: src/js/lib/logging/test/test.js
Comment (s9ferech) [reply src/js/lib/logging/test/test.js:1]: To me it seems like this should better be included when you execute `./run test`, which the CI currently does. Am I right?

File: src/js/lib/content/src/index.js
Comment (s9ferech) [reply src/js/lib/content/src/index.js:149]: Even if the IDE runs into an infinite recursion and exceeds the stack size limit, I wouldn't expect more than a few MB of trace data, which still seems okay. But I could also just set the limit to 100 frames or so.

Comment (wdanilo) [reply src/js/lib/content/src/index.js:149]: I'd stay with 100 frames then. Think about a bug, which would cause massive crashes on many machines. If we've got reported 1k or 10k crashes, each containing several mb, this can be hard to handle.

File: src/js/lib/logging/README.md
Comment (wdanilo) [reply src/js/lib/logging/README.md:1]: We have several things like that in `lib` already, maybe we should rename `lib` to something else then. Basically icon generation is standalone exec as well, used in build process. The `content` lib is also a "stand-alone" web-app. The client is a standalone electron app as well. They are somehow connected (icons and content are bundled with other things during electron app building). Basically, the `lib` folder keeps all JS projects together now.

File: src/js/lib/logging/README.md
Comment (wdanilo) [reply src/js/lib/logging/README.md:3]: Hmm, few thoughts / questions here:
1. There are 2 other options to handle it - what do you think about it?
    1. Make `client.js` always pass the default parameter even if its not explicitly provided in the url.
    2. In case the URL does not contain the option, just disable log sending.

2. Regarding naming - I like it, lets do it.

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [reply src/js/lib/logging/test/test.js:1]: Oh, you are right. This way it would be much better. Thanks for catching it, I was definitely tired writing it ;)

**Review by wdanilo — COMMENTED** [review]
Sorry for the super late comments, TBH I missed your comments completely. In case I would ever miss them again, just ping me on Discord. I should review all PRs on daily basis, without making you or anyone wait. I apologize once again!

File: src/js/lib/logging/README.md
Comment (s9ferech) [reply src/js/lib/logging/README.md:3]: > Make `client.js` always pass the default parameter even if its not explicitly provided in the url.

This seems like a good solution. The best way would likely be to pass the address through a cookie then.

> In case the URL does not contain the option, just disable log sending.

This would be trivial to implement. But is it intended that users will also open the IDE in their browser? If so, then this does not seem optimal to me.

There would also be the possible solution to send logs to the server that hosts the IDE. That server could then pass it on if necessary. But I don't know how the IDE will be used when it is finished, so I don't know what solution would  be best.

File: src/js/lib/logging/test/test.js
Comment (s9ferech) [reply src/js/lib/logging/test/test.js:1]: Then it should better be implemented by someone who knows how the run script works. You just need to execute `npm test` on the log-server package.

File: src/js/lib/logging/README.md
Comment (wdanilo) [reply src/js/lib/logging/README.md:3]: > This seems like a good solution. The best way would likely be to pass the address through a cookie then.

Why it would be better than passing it as we pass it now, in url? TBH I'm not a web-guy, so I cant really compare these two solutions in my head now, so I'd love to know why cookies can be better here. The only thing that makes me instantly dislike cookies is that we would then need to display info in the app "that it uses cookies", which I would rather not want to display to the user :) 

> But is it intended that users will also open the IDE in their browser.

Opening in the browser was always meant to be a very debug thing and only for advanced users. Of course, I'm not talking about Enso Cloud, that we will host, but in this case we will pass the values in a special way (Im not sure how yet). On the other hand, the more I think about it, the more Im sure that you are right, that the browser experience should be close to electron experience as much as possible, so yep, lets not disable it then.

So yeah, I think the best way to progress with that is to use the common config js file that all needed scripts will use to refer to default values.

File: src/js/lib/logging/test/test.js
Comment (wdanilo) [reply src/js/lib/logging/test/test.js:1]: Sure, we will implement it in separate PR then!

File: src/js/lib/logging/README.md
Comment (s9ferech) [reply src/js/lib/logging/README.md:3]: The problem with passing it through the URL would only be that you need to enter it when you open the front end in your browser (if you don't use the default). Cookies are just one way for the IDE server to send the configuration in the background. They seem like the simplest solution given the implementation of your server.

But in my opinion, the current solution is fine for the moment and I would only change it when you know for sure that you need something different.

**Review by wdanilo — APPROVED** [review]
Sorry, it took me a while to do the last review. I have to admit that this code is amazing and the results are super cool. Really, really good job @s9ferech ❤️ ❤️ ❤️

**Pull request merged**
